design/ingressroute-design.md

```diff
@@ -0,0 +1,398 @@
+I, the owner of the virtual host, delegate to you the /foo prefix and everything under it, implies that you cannot operate on this vhost outside /foo, but it also means that whatever you do in /foo; host malware, leak k8s session keys, is something I'm trusting you not to do.
+
+## No ability to prevent route delegation
+
+In writing up this proposal an issue occured to me.
+It happens when Contour is operating in the, poorly specified, enforcing mode; where Contour only recognises root IngressRoutes in a set of authorised namespaces.
+(well, it happens regurdless of enforcement mode, but the argument is that if you do not turn on enforcement you are explicitly saying you trust you users, or you have another mechanism in place -- CI/CD -- to handle this).
+
+Imaging this scenario:
+```yaml
+apiVersion: contour.heptio.com/v1alpha1
+kind: IngressRoute
+metadata:
+  # this is www.google.com's root object
+  name: google
+  namespace: virtualhost-namespace
+spec:
+  virtualhost:
+    fqdn: www.google.com
+    tls:
+      # required, the name of a secret in the current namespace
+      secretName: google-tls
+      # other properties like cipher suites may be added later
+  routes:
+  - match: /mail
+    # delegate to the gmail service running in another namespace
+    delegate:
+      name: gmail
+      namespace: gmail
+    ...
+```
+And in the gmail namespace we have a IngressRoute vertex
+```yaml
+apiVersion: contour.heptio.com/v1alpha1
+kind: IngressRoute
+metadata:
+  name: gmail 
+  namespace: gmail
+spec:
+  routes:
+  - match: /mail
+    service:
+      name: gmail-tomcat-svc
+      port: 8000
+```
+That's all fine, the operators in the `gmail` namespace can't change the delegation in the `virtualhost-namespace` and the admins in the `virtualhost-namespace` feel confident that they control `www.google.com` and that gmail is being served over tls.
+
+However, at some time later ad money is getting tight and the operators take on a new customer, who is very cheap and doesn't want to pay for tls.
+```yaml
+apiVersion: contour.heptio.com/v1alpha1
+kind: IngressRoute
+metadata:
+  name: daves-cheap-webhosting
+  namespace: virtualhost-namespace
+spec:
+  virtualhost:
+    fqdn: dave.cheney.net
+  routes:
+  - match: /
+    delegate:
+      name: webapp
+      namespace: customer-dave
+```
+Because the ingress operators don't want to have to handle tickets for dave's cheap webhosting, they delegate `/` to dave's namespace and tell him if he has a problem getting it working he can ask for help on the forum.
+
+_However_, dave deploys this vertex ingressroute CRD.
+```yaml
+apiVersion: contour.heptio.com/v1alpha1
+kind: IngressRoute
+metadata:
+  name: webapp 
+  namespace: customer-dave
+spec:
+  routes:
+  - match: /mail
+    delegate:
+      name: gmail
+      namespace: gmail
+```
+And all of a sudden, `http://dave.cheney.net/mail` is serving up the full gmail application, sans TLS.
```

Comment:
But couldn't the "attacker" do this by running a proxy and forwarding connections regardless?

While this is bad, I'm not sure it is something that couldn't be done in other ways.

Reply:
I think this is bad; I can expose any route for any service running on a k8s cluster if I can guess its name (which I can do very quickly via the API server). This means we cannot do things like enforcing TLS over a whole route, or apply basic auth or other restrictions in a way that propagate downwards as the attacker may create a complimentary route that elides those restrictions.

The k8s analog is a container nominates named ports, but those are optional (I don't know why they aren't mandatory, but that's beside the point) as the service document can publish any port it wants. The mitigating control is service documents cannot span namespaces. In the design we're considering it is possible for me to breach the namespace boundary by delegating to a route in another IngressRoute's namespace.

I think some option to make delegation bi directional is needed to close this hole.